File: Python_Algorithm/brushfire.py
import argparse
import time
import tracemalloc
from collections import deque

from read_from_csv import read_from_csv
from write_to_json import write_to_json

def is_valid_neighbour_tile(row_index, column_index, map, algorithm_map):
    """Check if a given tile is part of the map and was not visited before.

    Arguments:
        row_index (int): The row index of the current position.
        column_index (int): The columns index of the current position.
        map (list[list[int]]): The map on which the Wavefront-algorithm is applied.
        algorithm_map (list[list[int]]): The current map written by the Wavefront-algorithm.

    Returns:
        boolean: Returns whether the tile is a valid neighbour or not.
    """
    return (
        0 <= row_index < len(map)
        and 0 <= column_index < len(map[0])
        and algorithm_map[row_index][column_index] == -1
    )


def brushfire(map, algorithm_map, status_code):
    """Apply the Brushfire Algorithm.

    Arguments:
        map (list[list[int]]): The map in form of a list of lists of integers.
        algorithm_map (list[list[int]]): The prepared map in form of a list of lists of integers that will be overwritten by the Brushfire-algorithm.
        status_code (int): The status code returned when reading the CSV-file.

    Returns:
        list[list[int]]: The map created by the Brushfire-algorithm filled with the determined values.
        int: The status code symbolizing success or a specific error type prior to or while performing the Brushfire-algorithm.
        int: The shortest determined path length in the form of tiles towards the goal.
        list[tuple[int, int]]: The shortest determined path in form of the positions towards the goal.
        float: The time the Brushfire-algorithm needs to finish in seconds.
        float: The memory used during the execution in MB.
    """
    # Checks if an error was encountered prior to the Brushfire-algorithm
    if status_code != 200:
        return algorithm_map, status_code, -1, [], 0.0, 0.0

    # Saves the start_time and the observation of the memory usage is started
    start_time = time.perf_counter()
    tracemalloc.start()

    # Initializes all the components needed for the Brushfire-algorithm
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    position_queue = deque()
    path_length = -1
    path_positions = []

    # Initialize distances for obstacles
    for row_index in range(len(map)):
        for column_index in range(len(map[0])):
            if map[row_index][column_index] == 1:
                algorithm_map[row_index][column_index] = 0
                position_queue.append((row_index, column_index, 0))

    # Performs the Brushfire-algorithm as long as there are still tiles in the position_queue with unchecked neighbours
    while position_queue:
        # Reads a position and saved distance from the position_queue according to the First-In-First-Out principle
        row_index, column_index, current_distance = position_queue.popleft()

        # For each of the four tiles next to the read position the distance towards an obstacle is saved
        for row_direction, column_direction in directions:
            current_row_index = row_index + row_direction
            current_column_index = column_index + column_direction

            # Prior to saving the distance it is checked, if the tile is a valid neighbour
            if is_valid_neighbour_tile(current_row_index, current_column_index, map, algorithm_map):
                # The determined distance gets saved to the algorithm_map and the field gets added to the position_queue
                algorithm_map[current_row_index][current_column_index] = current_distance + 1
                position_queue.append((current_row_index, current_column_index, current_distance + 1))

    # The memory usage gets saved and the tracking is stopped
    _, peak = tracemalloc.get_traced_memory()
    tracemalloc.stop()

    # The end_time gets saved and the computing_time in seconds is calculated by subtracting the start_time and rounding the result to two decimal places
    end_time = time.perf_counter()
    computing_time = end_time - start_time

    # The memory usage gets converted to MB
    memory_usage = peak / 10**6

    return algorithm_map, status_code, path_length, path_positions, computing_time, memory_usage


if __name__ == "__main__":
    """The main function of the Brushfire-algorithm.

    Arguments:
        csv_path (string): The path towards the CSV-file with the map the Brushfire-algorithm is supposed to be applied to.
        json_path (string): The path towards the JSON-file where the results of the Brushfire-algorithm are saved.
    """
    parser = argparse.ArgumentParser(description="Brushfire-Algorithmus auf Basis einer Map in einer CSV-Datei und Ausgabe in einer JSON-Datei")
    parser.add_argument("csv_path", help="Pfad zur Eingabe-CSV-Datei")
    parser.add_argument("json_path", help="Pfad zur Ausgabe-JSON-Datei")
    args = parser.parse_args()

    # Reads in the map, applies the Brushfire-algorithm to it and saves the results to a JSON-file
    map, algorithm_map, _, _, status_code = read_from_csv(args.csv_path)
    algorithm_map, status_code, path_length, path_positions, computing_time, memory_usage = brushfire(map, algorithm_map, status_code)
    write_to_json(args.json_path, algorithm_map, status_code, path_length, path_positions, computing_time, memory_usage)

File: Java_Algorithms/Brushfire.java
import java.lang.management.ManagementFactory;
import java.lang.management.MemoryPoolMXBean;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Brushfire is the class that applies the Brushfire-algorithm
 */
public class Brushfire {
    /**
     * The four directions to neighbour tiles
     */
    private static final int[][] DIRECTIONS = {
            {-1, 0}, {1, 0}, {0, -1}, {0, 1}
    };

    /**
     * The map in form of a multidimensional array
     */
    private final int[][] map;

    /**
     * The prepared map in form of a multidimensional array that will be overwritten by the Brushfire-algorithm
     */
    private final int[][] algorithmMap;

    /**
     * The status code returned when reading the CSV-file
     */
    private final int statusCode;

    /**
     * Constructor of the Brushfire class
     * @param map The map in form of a multidimensional array
     * @param algorithmMap The prepared map in form of a multidimensional array that will be overwritten by the Brushfire-algorithm
     * @param statusCode The status code returned when reading the CSV-file
     */
    public Brushfire (int[][] map, int[][] algorithmMap, int statusCode) {
        this.map = map;
        this.algorithmMap = algorithmMap;
        this.statusCode = statusCode;
    }

    /**
     * Check if a given tile is part of the map, not an obstacle, and was not visited before
     * @param rowIndex The row index of the current position
     * @param columnIndex The columns index of the current position
     * @return Whether the tile is a valid neighbour or not
     */
    private boolean isValidNeighbourTile(int rowIndex, int columnIndex) {
        return (
                (0 <= rowIndex && rowIndex < map.length) &&
                        (0 <= columnIndex && columnIndex < map.length) &&
                        (algorithmMap[rowIndex][columnIndex] == -1)
        );
    }

    /**
     * The actual Brushfire-algorithm
     * @return The results of the algorithm in form of an AlgorithmResult
     */
    public AlgorithmResult applyBrushfire() {
        // Checks if an error was encountered prior to the Brushfire-algorithm
        if (statusCode != 200) {
            return new AlgorithmResult(algorithmMap, statusCode, -1, new int[][]{}, 0, 0);
        }

        // Saves the startTime and starts the observation of the memory usage
        long memoryBefore = 0;

        for (MemoryPoolMXBean memoryPool : ManagementFactory.getMemoryPoolMXBeans()) {
            long usedMemory = memoryPool.getUsage().getUsed();
            if (usedMemory > 0) {
                memoryBefore = memoryBefore + usedMemory;
            }
        }

        long startTime = System.nanoTime();

        // Initializes all the components needed for the Brushfire-algorithm
        Deque<int[]> positionQueue = new ArrayDeque<>();
        int pathLength = -1;
        int[][] pathPositions = new int[][]{};

        // Initialize distances for obstacles
        for (int rowIndex = 0; rowIndex < map.length; rowIndex++) {
            for (int columnIndex = 0; columnIndex < map[rowIndex].length; columnIndex++) {
                if (map[rowIndex][columnIndex] == 1) {
                    algorithmMap[rowIndex][columnIndex] = 0;
                    positionQueue.addLast(new int[]{rowIndex, columnIndex, 0});
                }
            }
        }

        // Performs the Brushfire-algorithm as long as there are still tiles in the positionQueue with unchecked neighbours
        while (!positionQueue.isEmpty()) {
            // Reads a position and saved distance from the position_queue according to the First-In-First-Out principle
            int[] position = positionQueue.removeFirst();
            int rowIndex = position[0];
            int columnIndex = position[1];
            int distance = position[2];

            // For each of the four tiles next to the read position the distance towards an obstacle is saved
            for (int[] direction : DIRECTIONS) {
                int currentRowIndex = rowIndex + direction[0];
                int currentColumnIndex = columnIndex + direction[1];

                // Prior to saving the distance it is checked, if the tile is a valid neighbour
                if (isValidNeighbourTile(currentRowIndex, currentColumnIndex)) {
                    // The determined distance gets saved to the algorithmMap and the field gets added to the positionQueue
                    algorithmMap[currentRowIndex][currentColumnIndex] = distance + 1;
                    positionQueue.addLast(new int[]{currentRowIndex, currentColumnIndex, distance + 1});
                }
            }
        }

        // The endTime gets saved
        long endTime = System.nanoTime();

        // The tracking of the memory usage is stopped
        long memoryAfter = 0;

        for (MemoryPoolMXBean memoryPool : ManagementFactory.getMemoryPoolMXBeans()) {
            long usedMemory = memoryPool.getUsage().getUsed();
            if (usedMemory > 0) {
                memoryAfter = memoryAfter + usedMemory;
            }
        }

        // The computingTime in milliseconds is calculated by subtracting the startTime
        double computingTime = (double)(endTime - startTime)/1_000_000.0;

        // The memory usage gets converted to MB
        double memoryUsage = (double)(memoryAfter - memoryBefore) / 1_048_576.0;

        BigDecimal decimalComputingTime = new BigDecimal(computingTime).setScale(19, RoundingMode.HALF_UP);

        return new AlgorithmResult(algorithmMap, statusCode, pathLength, pathPositions, decimalComputingTime, memoryUsage);
    }
}
